### ISSUE TITLE
v2.0.0 breaking changes

### ISSUE THREAD

**Issue by koresar** [ericelliott/stampit#43]
I have just created the branch [`v2_0`](https://github.com/ericelliott/stampit/compare/v2_0). It contains fixes for #40 which break compatibility with stampit 1.x.
But I am not creating a PR yet because I believe there are few more breaking changes necessary.

## Immutable stamps
Let's make **all** chain methods to return new stamp. Let's make stamps immutable after it's created.

### A long story.
When I first started using stampit I thought that `stamp.state`, `stamp.methods`, and `stamp.enclose` create new stamps from existing. I was wrong eventually.
Today @raabbajam showed me his code snippets. He assumed the same!

### Why it is good.
1. This is what people expect by default.
1. This will make all chaining methods consistent (with `stamp.compose`).
1. In some cases the code would be much shorter: `bigDog = dog.state({ size: 'big' })` instead of `bigDog = dog.compose(stampit(null, { size: 'big' }))`.
1. (IMO) Changing existing stamp is useless feature. Nobody need changes to the `this` of his stamps.
1. And a mini bonus. All stamps are currently referencing entire stampit closure (methods like extractFunction, slice, forEach, forOwn. 10 in total). We could avoid that!

## ~~`combine` instead of `compose`~~ CANCELLED
I browsed few English dictionaries. I believe the better wording for the `compose` function would be `combine`.

@ericelliott I would really love to hear your opinion on these two.

**Comment by ericelliott** [comment]
> Immutable stamps

Yeah! I've been thinking I should have done this from the start for about two years. I secretly had this in mind for a 2.0 release. *It's like you're reading my mind!*

> Today @raabbajam showed me his code snippets. He assumed the same!

This is a common question I hear, which is what got me thinking immutability should have been the plan in the first place.

The **only** reason I didn't do this was because I had an edge case that required a mutable stamp to work, but looking back, I think that the decision to bend my ideal API to fit *one* use case that I've never had again was an error.


## combine vs / compose

The reason I chose compose is rooted in the famous GoF book, "Design Patterns", where they state two foundational principles on which the book was built. One of them being, "Favor object composition over class inheritance."

In other words, I specifically chose the word to:

1. Emphasize that `.compose()` does not suffer the same problems as class inheritance.
2. Use a word that should be well-known to the OOP community.

**Comment by koresar** [comment]
Re combine/compose.

"Favor object composition over class inheritance." The GoF reference **objects**, not classes/stamp/factories. IMO you we should _compose_ objects, but _combine_ stamps.

I insist on voting among stampit community. :)

**Comment by koresar** [comment]
Re the immutability. Let me create a PR for the v2_0 branch.

**Comment by ericelliott** [comment]
OK. If you can get ten other people to vote for combine, we'll add it as an alias. Get 20 and we'll deprecate `.compose` (unless we hear lots of objections).

**Comment by RainerAtSpirit** [comment]
@ericelliott : You asked for it :). +1 for `combine` stamps

**Comment by koresar** [comment]
### Guys, there is the poll for combine/compose.
I encourage you (and other stampit users you know) to vote.
http://goo.gl/forms/O3yL2OYQsy

**Comment by ericelliott** [comment]
:)

**Comment by StevenPerez** [comment]
👍🎉🎊

**Comment by foxaal** [comment]
If you start hunting around Wikipedia regarding the terms "compose" and "combine", you quickly come to the realization that "compose" makes more sense against the past.  The term "aggregate" also come up, but "combine" seems to be nowhere.  Is the objection to "compose" that it might get confused with matrix composition?  When does the voting stop, or has it already?

**Comment by ericelliott** [comment]
Let's wait for resolution on which name will replace `.compose()` before we merge this work. =)

See https://github.com/ericelliott/stampit/issues/46

**Comment by koresar** [comment]
I didn't add `combine` alias with any of the existing PR's as I was waiting for the poll results. The alias(-es) should be added with a separate PR I reckon.

**Comment by ericelliott** [comment]
I can see using a separate commit, but I don't think a separate PR is warranted. The 2.0 API should be fleshed out when we merge it. There will be no more breaking changes in the 2.0 branch after it is merged. Otherwise, we'd have to increment to 3.0 right away in order to comply with [semver](http://semver.org/).

**Comment by koresar** [comment]
See #48. I am waiting you to review the `mixInChain` to `v2_0` merge. `master` branch will stay untouched if you accept the #48.

**Comment by koresar** [comment]
For future references.

Re the **Compose/Combine** dilemma. It was decided to leave the naming as is. No new aliases, no changes.

**Comment by ericelliott** [comment]
=)

**Comment by koresar** [comment]
Immutability is done in branch `v2_0`.
